service/worker/migration/activities.go

```diff
@@ -346,3 +348,53 @@ func (a *activities) GenerateReplicationTasks(ctx context.Context, request *gene
 
 	return nil
 }
+
+func (a *activities) SeedReplicationQueueWithUserDataEntries(ctx context.Context, params TaskQueueUserDataReplicationParamsWithNamespace) error {
+	if len(params.Namespace) == 0 {
+		return temporal.NewNonRetryableApplicationError("namespace is required", "InvalidArgument", nil)
+	}
+	if params.PageSize == 0 {
+		params.PageSize = defaultPageSizeForTaskQueueUserDataReplication
+	}
+	if params.RPS == 0 {
+		params.RPS = defaultRPSForTaskQueueUserDataReplication
+	}
+
+	describeResponse, err := a.frontendClient.DescribeNamespace(ctx, &workflowservice.DescribeNamespaceRequest{
+		Namespace: params.Namespace,
+	})
+	if err != nil {
+		return err
+	}
+
+	rateLimiter := quotas.NewRateLimiter(params.RPS, int(math.Ceil(params.RPS)))
+	var nextPageToken []byte
+
+	if activity.HasHeartbeatDetails(ctx) {
+		if err := activity.GetHeartbeatDetails(ctx, &nextPageToken); err != nil {
+			return temporal.NewNonRetryableApplicationError("failed to load previous heartbeat details", "TypeError", err)
+		}
+	}
+
+	for {
+		if err := rateLimiter.Wait(ctx); err != nil {
+			return err
+		}
+
+		request := &matchingservice.SeedReplicationQueueWithUserDataEntriesRequest{
+			NamespaceId:   describeResponse.GetNamespaceInfo().Id,
+			NextPageToken: nextPageToken,
+			PageSize:      int32(params.PageSize),
+		}
+		response, err := a.matchingClient.SeedReplicationQueueWithUserDataEntries(ctx, request)
```

Comment:
why push this work to some arbitrary matching node? why not just do it here?

Reply:
each individual matching node owns the user data for task queues that are currently assigned to it by membership. but in this case you're having one node read data owned by other nodes, so you're already breaking the ownership model. which is totally fine because you're just reading. but then there's no reason not to just do it here. the fact that you have to use `"not-applicable"`/`UNSPECIFIED` is a clue that it's not a good fit.

from an operational/scaling pov, I'd also rather do as much migration work as possible on the workers, to reduce the impact on the rest of the system for background tasks. (of course the writes have to go through matching because of ownership)

Reply:
I had a feeling you'd say that, thought of that too.
I'm fine with moving the functionality into the worker.

Reply:
It's making this harder to test, because now my XDC integration test needs direct access to the matching DB.
I vote for keeping it the way it is unless you feel _that_ strongly against it.

Reply:
I do feel pretty strongly but we can talk about it.

I think the boilerplate required to get a TaskManager in the integration test is less than the boilerplate to add an rpc to matching, right?

Reply:
I'm convinced, PR coming in a bit.